File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        int l=1,r=1000000,ans=-1,mid;
        long long total = 0;
        while(l<r){
            mid = l+(r-l)/2;
            total=0;
            for(auto& i:nums) total+=((i/mid) + (i%mid?1:0));

            if(total>threshold) l=mid+1;
            else {
                // ans=mid;
                r=mid;
            }
        }
        return l;
    }
};

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.py
from math import floor,ceil 

class Solution:
    def getSumDivisions(self,nums,divisor): 
        total = 0
        for val in nums: 
            total+=ceil(val/divisor) 
        return total 


    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        #1. range -  [1,max(nums)] all possible 
        #2. sum - all the divisons and result should be greater than the threshold 
        l,r=1,max(nums)
        res=-1 
        while l<=r: 
            mid=l+(r-l)//2
            if self.getSumDivisions(nums,mid)<=threshold: 
                res=mid
                r=mid-1
            else: 
                l=mid+1 

        return res 